segmentstore/storage/impl/src/main/java/io/pravega/segmentstore/storage/impl/filesystem/FileSystemStorage.java

```diff
@@ -0,0 +1,407 @@
+        }
+
+        long fileSize = path.toFile().length();
+        if (fileSize < offset) {
+            throw new BadOffsetException(handle.getSegmentName(), fileSize, offset);
+        } else {
+            try (FileChannel channel = FileChannel.open(path, StandardOpenOption.WRITE);
+                 ReadableByteChannel sourceChannel = Channels.newChannel(data)) {
+                while (length != 0) {
+                    long bytesWritten = channel.transferFrom(sourceChannel, offset, length);
+                    offset += bytesWritten;
+                    length -= bytesWritten;
+                }
+            }
+            LoggerHelpers.traceLeave(log, "write", traceId);
+            return null;
+        }
+    }
+
+    private boolean isWritableFile(Path path) throws IOException {
+        PosixFileAttributes attrs = Files.readAttributes(path, PosixFileAttributes.class);
+        return attrs.permissions().contains(OWNER_WRITE);
+    }
+
+    @SneakyThrows
+    private Void syncSeal(SegmentHandle handle) {
+        long traceId = LoggerHelpers.traceEnter(log, "seal", handle.getSegmentName());
+        if (handle.isReadOnly()) {
+            throw new IllegalArgumentException(handle.getSegmentName());
+        }
+
+        Set<PosixFilePermission> perms = new HashSet<>();
+        // add permission as r--r--r-- 444
+        perms.add(PosixFilePermission.OWNER_READ);
+        perms.add(PosixFilePermission.GROUP_READ);
+        perms.add(PosixFilePermission.OTHERS_READ);
+        Files.setPosixFilePermissions(Paths.get(config.getFilesystemRoot(), handle.getSegmentName()), perms);
+        LoggerHelpers.traceLeave(log, "seal", traceId);
+        return null;
+    }
+
+    @SneakyThrows(IOException.class)
+    private Void syncConcat(SegmentHandle targetHandle, long offset, String sourceSegment) {
+        long traceId = LoggerHelpers.traceEnter(log, "concat", targetHandle.getSegmentName(),
+                offset, sourceSegment);
+
+        Path sourcePath = Paths.get(config.getFilesystemRoot(), sourceSegment);
+        Path targetPath = Paths.get(config.getFilesystemRoot(), targetHandle.getSegmentName());
+
+        try (FileChannel targetChannel = (FileChannel) Files.newByteChannel(targetPath, EnumSet.of(StandardOpenOption.APPEND));
+             RandomAccessFile sourceFile = new RandomAccessFile(String.valueOf(sourcePath), "r")) {
+            if (isWritableFile(sourcePath)) {
+                throw new IllegalStateException( String.format("Source segment (%s) is not sealed.", sourceSegment));
+            }
+            long length = sourceFile.length();
+            while ( length > 0 ) {
+                long bytesTransferred = targetChannel.transferFrom(sourceFile.getChannel(),
+                        offset, length);
+                offset += bytesTransferred;
+                length -= bytesTransferred;
+            }
+            Files.delete(sourcePath);
```

Comment:
If this happens, then the StorageWriter will reattempt the operation later on (since it never succeeded). You need to make sure your code handles that and that it won't throw some weird, unexpected exception otherwise.